File: WpfApp1/TcpProtocol/SendImage.cs

using System.Diagnostics;

namespace TcpProtocol
{
    public class SendImageQuery : Query
    {
        public int BufferWidth { get; set; }
        public int BufferHeight { get; set; }
        public int Width { get; set; }
        public int Height { get; set; }
        public byte[] ImageBuffer { get; set; }

        public SendImageQuery()
        {
            BufferWidth = 0;
            BufferHeight = 0;
            Width = 0;
            Height = 0;
            ImageBuffer = null;
        }

        public SendImageQuery(int bufferWidth, int bufferHeight, int width, int height, byte[] imageBuffer)
        {
            BufferWidth = bufferWidth;
            BufferHeight = bufferHeight;
            Width = width;
            Height = height;
            ImageBuffer = imageBuffer;
        }

        public override void Serialize(Serializer serializer)
        {
            serializer.WriteInt(BufferWidth);
            serializer.WriteInt(BufferHeight);
            serializer.WriteInt(Width);
            serializer.WriteInt(Height);
            serializer.WriteBytes(ImageBuffer);
        }

        public override void Deserialize(Serializer serializer)
        {
            BufferWidth = serializer.ReadInt();
            BufferHeight = serializer.ReadInt();
            Width = serializer.ReadInt();
            Height = serializer.ReadInt();
            ImageBuffer = serializer.ReadBytes();
        }

        public override Response Execute()
        {
            return new SendImageResponse();
        }
    }

    public class SendImageResponse : Response
    {
        public SendImageResponse()
        {

        }

        public override void Serialize(Serializer serializer)
        {

        }

        public override void Deserialize(Serializer serializer)
        {

        }
    }
}

File: WpfApp1/Runtime/TcpProtocol/SendImage.cpp

#include "SendImage.h"


namespace TcpProtocol
{
    // Query
    bool SendImageQuery::Serialize(Serializer& serializer) const
    {
        if (!serializer.WriteInt(width)) { return false; }
        if (!serializer.WriteInt(height)) { return false; }
        if (!serializer.WriteBuffer(imageBuffer)) { return false; }
        return true;
    }

    bool SendImageQuery::Deserialize(Serializer& serializer)
    {
        if (!serializer.ReadInt(width)) { return false; }
        if (!serializer.ReadInt(height)) { return false; }
        if (!serializer.ReadBuffer(imageBuffer)) { return false; }
        return true;
    }

    const char* SendImageQuery::Name() const
    {
        return "TcpProtocol.SendImageQuery";
    }

    std::unique_ptr<Response> SendImageQuery::Execute()
    {
        return std::make_unique<SendImageResponse>();
    }


    // Response
    bool SendImageResponse::Serialize(Serializer&) const
    {
        return true;
    }

    bool SendImageResponse::Deserialize(Serializer&)
    {
        return true;
    }

    const char* SendImageResponse::Name() const
    {
        return "TcpProtocol.SendImageResponse";
    }
}

